Adding Visual Test Report in Github Actions

Vaivaswat2244 opened the pull request:
<!--
  Thank you for contributing! Please use this pull request (PR) template.


 In the description field of this PR, include "resolves #XXXX" tagging the issue you are fixing. If this PR addresses the issue but doesn't completely resolve it (ie the issue should remain open after your PR is merged), write "addresses #XXXX".-->
Resolves #7618 

This PR adds support for generating an HTML report to compare visual test results more effectively. The goal is to improve accessibility and readability of visual regression test outputs.

 Changes:
<!-- Add here what changes were made in this pull request and if possible provide links showcasing the changes. -->
* Added a script to generate an HTML report summarizing visual test results.
* The report includes expected screenshots and actual screenshots
* Updated the workflow to run the added script after completion of tests
* Edited the VisualTest.js to save the recorded screenshots to a folder actual-screenshots.
* Then the generated html report is uploaded as an artifact.


 Screenshots of the change:
This is the screenshot of the html page generated. It will show expected and Actual visuals of this.
<!-- If applicable, add screenshots depicting the changes. -->
![image](https://github.com/user-attachments/assets/7123a635-d63b-4932-b948-ce2363b3d090)


#### PR Checklist
<!--
  To check any option, replace the "[ ]" with a "[x]". Be sure to check out how it looks in the Preview tab! Feel free to remove any portion of the template that is not relevant for your issue.
-->

- [x] `npm run lint` passes
- [ ] [Inline reference] is included / updated
- [ ] [Unit tests] are included / updated

[Inline reference]: https://p5js.org/contribute/contributing_to_the_p5js_reference/
[Unit tests]: https://github.com/processing/p5.js/tree/main/contributor_docs#unit-tests

ksen0 commented:
Thanks for your patience @Vaivaswat2244 ! I will have time next week (when the GSoC reviews are done:) to review this, though if someone else reviews this before their feedback is welcome. Sorry for the wait.

Vaivaswat2244 commented:
Hey @ksen0 , can this be reviewed now?

ksen0 commented:
Hi @Vaivaswat2244 thanks for all the patience on this and for the reminder. It's really nicely done! I've added notes on 3 points below, let me know what you think about all that, if you have any other ideas (or any questions). 

The report looks great, especially the filter for "show only failed."

(1) There are false positives for FAIL - these tests passed but are missing images - could you investigate why this happens please?

<img width="1235" alt="image" src="https://github.com/user-attachments/assets/fd64d4f8-7422-4084-a23e-e2b649dae7e4" />

(2) Very, very minor nice-to-haves - for consistency, use lowercase step names ("upload visual test report"). If it's easy to add, include link to PR or the # in the report html. Feel free to ignore this feedback.

(3) Major - execution conditions:

<img width="791" alt="image" src="https://github.com/user-attachments/assets/8e384022-f491-493c-ae42-e049a67e9109" />

So if there is a failing case, then `npm test` fails and report is not generated, preventing it from being used. This is the desired outcome when tests fail:

<img width="225" alt="image" src="https://github.com/user-attachments/assets/b6f9e50f-fae1-41ba-ae94-59e80351c641" />

In the example above I'm using:

```yml
name: Testing

on:
  push:
    branches:
      - main
      - dev-2.0
  pull_request:
    branches:
      - '*'

jobs:
  test:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v1
      - name: Use Node.js 20.x
        uses: actions/setup-node@v1
        with:
          node-version: 20.x
      - name: Get node modules
        run: npm ci
        env:
          CI: true
      - name: build and test
        id: test
        run: npm test
        continue-on-error: true
        env:
          CI: true
      - name: Generate Visual Test Report
        if: always()
        run: node visual-report.js
        env:
          CI: true
      - name: Upload Visual Test Report
        if: always()
        uses: actions/upload-artifact@v4
        with:
          name: visual-test-report
          path: test/unit/visual/visual-report.html
          retention-days: 14
      - name: report test coverage
        if: steps.test.outcome == 'success'
        run: bash <(curl -s https://codecov.io/bash) -f coverage/coverage-final.json
        env:
          CI: true
      - name: fail job if tests failed
        if: steps.test.outcome != 'success'
        run: exit 1
```


I was able to test this by adding an artificial problem into the test suite:

```js
    visualTest("with the default font", function (p5, screenshot) {
      p5.createCanvas(50, 50);
      p5.textSize(20);
      p5.textAlign(p5.LEFT, p5.TOP);
      p5.text("wrong text", 0, 0); // should be 'text' so should defintiely fail
      screenshot();
    });
```

This failure does show up in the report, but without actual image:

<img width="1223" alt="image" src="https://github.com/user-attachments/assets/b1a2f5c5-f836-49cc-a926-918732461347" />

Looks like `writeImageFile(actualFilename, toBase64(actual[i]));` only runs on success, so in addition to updating the ci-test.yml, please check the logic of the visualTest.js. The goal is to ensure that **report should help users see the visual test failures.**

If problem (3) is addressed, happy to merge!

Vaivaswat2244 commented:
Got it! I reviewed the code and yes currently the report is not generated in case of test failure. This is a logical misunderstanding from my side which I'll fix. Although regarding the false positives I'm not sure, I'll test the new logic and most probably that'll fix the (1) and (3) points. Also I'll update the script for adding diff as well. Thanks for the detailed review on the pr...

Vaivaswat2244 commented:
hey @ksen0 @davepagurek,
I've tried  to fix the above mentioned issues. I've checked the logic of visualTest.js and it now writes actual-screenshots no matter what...
Also I've updated the ci-test.yml file as @ksen0 recommended.
Although I'm stuck on the false fail issue... 
This one test "Drawing triangles" passes on all the tests. Still the actual screenshot is not getting written in the actual-screenshots folder, when I've changed the logic and it is working for "intentionally failing test". I'm continuing to fix that. In the meantime I've commited fix for (2), (3) and (4) for review.

Vaivaswat2244 commented on test/unit/visual/visualTest.js:
```diff
@@ -452,6 +458,7 @@ export function visualTest(
           }
         } else {
           writeImageFile(expectedFilenames[i], toBase64(actual[i]));
+          writeImageFile(actualFilename, toBase64(actual[i]));
```
this was written because earlier the command was written in both if and else conditions. Now, I've moved `writeImageFile(actualFilename, toBase64(actual[i]));` outside the conditions

Vaivaswat2244 commented on .gitignore:
```diff
@@ -23,4 +23,6 @@ yarn.lock
-__screenshots__/
\ No newline at end of file
+__screenshots__/
+actual-screenshots/
```
Well, the `__screenshots__` here are not of the visual tests as they are not in the .gitignore file and are present in the repo. 
Though I will change the name of the folder to screenshots_actual if you wish.

Vaivaswat2244 commented:
Hey guys, @ksen0 @davepagurek. 
It's been some time since I looked into this. I've been thinking of adding a similar reporting system in the processing4 repository as the basis has been set up now (not pushed though). I'd love to hear any thoughts on this, whether this is appropriate, or if we can think of something better for the reporting systems of the visual testing aspects for both repositories.

Currently in the processing's system the initial structure of the visual testing system is set and working [have a look](https://github.com/Vaivaswat2244/processing4/tree/pr05-visualtests/visual-tests). I'm currently working towards adding more tests to the system. In the meantime, I was hoping we could resolve this to get better insights into the reporting system.
